### PR TITLE
Round Robin Group Assignment

### PR DESCRIPTION
**Pull request by nateh777**
##### What does the change add or fix?

Created new Rock job to be able to select people from a dataview and distribute them into a group structure based on campus on the person and the group. Primarily for use for deacons to be assigned to members automatically and balanced.

**New Settings:**

- People to Add Data View
- Groups to Cycle Through
- Default Campus
- Default Group

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

Created new Rock job to be able to select people from a dataview and distribute them into a group structure based on campus on the person and the group.

---------

##### Who reported, requested, or paid for the change?

RSC

---------

##### Does this update or add options to the block UI?

image

---------

##### What files does it affect?

- rocks.kfs.GroupRoundRobinAssignment/Jobs/GroupRoundRobinAssignment.cs
- rocks.kfs.GroupRoundRobinAssignment/Properties/AssemblyInfo.cs
- rocks.kfs.GroupRoundRobinAssignment/Properties/rocks.kfs.GroupRoundRobinAssignment.csproj
- rocks.kfs.GroupRoundRobinAssignment/rocks.kfs.GroupRoundRobinAssignment.csproj
- 

---------

##### Is it a breaking change for other versions/clients?

No

### CODE REVIEW

File: rocks.kfs.GroupRoundRobinAssignment/Jobs/GroupRoundRobinAssignment.cs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+	            CONCAT(pg.Name,' > ', g.Name)
+            ELSE
+                g.Name 
+        END",
+        IsRequired = true,
+        Key = AttributeKey.GroupsToCycleThrough )]
+
+    [CampusField(
+        name: "Default Campus",
+        description: "Default campus to assign people to and use if a family does not have a campus. If both group and campus are set, group takes precedence. If neither are set the person will remain unassigned.",
```
Comment (nateh777) [inline rocks.kfs.GroupRoundRobinAssignment/Jobs/GroupRoundRobinAssignment.cs:76]:
```suggestion
        description: "Default campus to assign people to and use if a family does not have a campus. If both default group and default campus are set, default group takes precedence. If neither are set the person will remain unassigned.",
```

File: rocks.kfs.GroupRoundRobinAssignment/Jobs/GroupRoundRobinAssignment.cs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        required: false,
+        includeInactive: false,
+        key: AttributeKey.DefaultCampus )]
+
+    [GroupField(
+        "Default Group",
+        Description = "Default group to assign people to and use if a family does not have a campus. If both group and campus are set, group takes precedence. If neither are set the person will remain unassigned.",
```
Comment (nateh777) [inline rocks.kfs.GroupRoundRobinAssignment/Jobs/GroupRoundRobinAssignment.cs:83]:
```suggestion
        Description = "Default group to assign people to and use if a family does not have a campus. If both default group and default campus are set, default group takes precedence. If neither are set the person will remain unassigned.",
```

File: rocks.kfs.GroupRoundRobinAssignment/Jobs/GroupRoundRobinAssignment.cs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+        /// </summary>
+        public GroupRoundRobinAssignment()
+        {
+        }
+
+        /// <summary>
+        /// Executes the specified context.
+        /// </summary>
+        /// <param name="context">The context.</param>
+        public virtual void Execute(IJobExecutionContext context)
```
Comment (nateh777) [inline rocks.kfs.GroupRoundRobinAssignment/Jobs/GroupRoundRobinAssignment.cs:114]:
```suggestion
        public virtual void Execute( IJobExecutionContext context )
```

File: rocks.kfs.GroupRoundRobinAssignment/Jobs/GroupRoundRobinAssignment.cs [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+                        .Queryable( false, false ).AsNoTracking()
+                        .Where( paramExpression, whereExpression, null );
+
+                    // Get Groups and child groups from job setting.
+                    var groupQry = groupService.GetByIds( groupIds );
+                    var groups = new List<Group>();
+
+                    foreach ( var group in groupQry )
+                    {
+                        var childGroups = groupService.GetChildren( group.Id, 0, false, new List<int>(), new List<int>(), false, false );
```
Comment (nateh777) [inline rocks.kfs.GroupRoundRobinAssignment/Jobs/GroupRoundRobinAssignment.cs:161]: This is recursive, right?

File: rocks.kfs.GroupRoundRobinAssignment/Jobs/GroupRoundRobinAssignment.cs [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+                                GroupRoleId = groupToAddTo.GroupType.DefaultGroupRoleId.Value,
+                                GroupMemberStatus = GroupMemberStatus.Active
+                            };
+
+                            if ( groupMember.IsValidGroupMember( rockContext ) )
+                            {
+                                groupMemberService.Add( groupMember );
+                            }
+
+                            addedCount++;
```
Comment (nateh777) [inline rocks.kfs.GroupRoundRobinAssignment/Jobs/GroupRoundRobinAssignment.cs:221]:
```suggestion
                                groupMemberService.Add( groupMember );
                                addedCount++;
                            }
```

File: rocks.kfs.GroupRoundRobinAssignment/Jobs/GroupRoundRobinAssignment.cs [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+                        }
+                        else
+                        {
+                            errorMessages.Add( string.Format( "No Group found to add {0} ({1}) to. Please check the person's campus, the default settings on this job and group campuses to ensure there is a match available.", person.FullName, person.Id ) );
```
Comment (nateh777) [inline rocks.kfs.GroupRoundRobinAssignment/Jobs/GroupRoundRobinAssignment.cs:225]: Add more verbose messaging here based on campus or group campus does not match.

File: rocks.kfs.GroupRoundRobinAssignment/Properties/rocks.kfs.GroupRoundRobinAssignment.csproj [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>bin\Release\</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+    <Prefer32Bit>false</Prefer32Bit>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="DDay.iCal">
```
Comment (nateh777) [inline rocks.kfs.GroupRoundRobinAssignment/Properties/rocks.kfs.GroupRoundRobinAssignment.csproj:36]: Check for unused references and remove.

File: rocks.kfs.GroupRoundRobinAssignment/rocks.kfs.GroupRoundRobinAssignment.csproj [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>bin\Release\</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+    <Prefer32Bit>false</Prefer32Bit>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="DDay.iCal">
```
Comment (nateh777) [inline rocks.kfs.GroupRoundRobinAssignment/rocks.kfs.GroupRoundRobinAssignment.csproj:36]: Remove unused references.

**Review by guymaness2 — APPROVED** [review]

File: rocks.kfs.GroupRoundRobinAssignment/Jobs/GroupRoundRobinAssignment.cs (hunk 4 above)
Comment (nateh777) [reply rocks.kfs.GroupRoundRobinAssignment/Jobs/GroupRoundRobinAssignment.cs]: No it is not, so I decided to change methodology here.

**Review by guymaness2 — APPROVED** [review]

**Pull request merged**